sktime/transformers/rocket.py

```diff
@@ -0,0 +1,178 @@
+from numba import njit, prange
+import numpy as np
+import pandas as pd
+
+from .base import BaseTransformer
+
+__author__ = "Angus Dempster"
+__all__ = ["ROCKET"]
+
+class Rocket(BaseTransformer):
+    """ROCKET
+
+    RandOm Convolutional KErnel Transform
+
+    @article{dempster_etal_2019,
+      author  = {Dempster, Angus and Petitjean, Fran\c{c}ois and Webb, Geoffrey I},
+      title   = {ROCKET: Exceptionally fast and accurate time series classification using random convolutional kernels},
+      year    = {2019},
+      journal = {arXiv:1910.13051}
+    }
+
+    Parameters
+    ----------
+    input_length : int, length of input time series
+    num_kernels  : int, number of random convolutional kernels
+    num_channels : int, number of channels for multivariate time series (default 1)
+    """
+
+    def __init__(self,  input_length, num_kernels, num_channels = 1):
+        self.input_length = input_length
+        self.num_kernels = num_kernels
+        self.num_channels = num_channels
+        self.kernels = _generate_kernels(input_length, num_kernels, num_channels)
+
+    def fit(self, X, y = None):
+        return self
+
+    def fit_transform(self, X, y = None):
+        return self.fit(X).transform(X)
+
+    def transform(self, X):
+        """Transforms input time series using random convolutional kernels.
+
+        Parameters
+        ----------
+        X : pandas DataFrame, input time series (sktime format)
+
+        Returns
+        -------
+        pandas DataFrame, transformed features
+        """
+        return pd.DataFrame(_apply_kernels(self._to_numpy(X), self.kernels))
+
+    @staticmethod
+    def _to_numpy(X, a = None, b = None):
```

Comment:
We do have a tabularisation function, which essentially does the same, how does your method represent multivariate time series data? I'd prefer if we move this function to the [`sktime.utils.data_container`](https://github.com/alan-turing-institute/sktime/blob/master/sktime/utils/data_container.py) file as we may want to add this functionality to the data container eventually.

Reply:
-oIt looks like `tabularize` 'flattens' the input into 2d.  This `_to_numpy` function produces an ndarray with three dimensions: (*number_of_examples*, *number_of_channels*, *length_of_time_series*).  (The `a` and `b` parameters are in order to be able to do the conversion in a rolling fashion in order to, in turn, avoid duplicating a big dataset in memory.)

This is basically the input format for our original implementation as well (extended to multivariate).  The problem, of course, comes with variable-length time series, for which I haven't really settled on a solution.  My favoured one is basically to mimic the backend of a scipy sparse matrix, store two sets of values: all the data in a single 1d numpy array, plus a set of lengths / cutpoints.

Reply:
Alright, let's keep it, I'd still prefer the `_to_numpy` method to go in the utils file. We've been reconsidering our data representation lately, let me know if you're interested, by the way are you in our slack?

Reply:
No worries, I'll move the function.

Definitely interested, if only to learn more (I don't pretend to know much about anything useful).  Not in the slack.

(I've been meaning to try [awkward-array](https://github.com/scikit-hep/awkward-array), as it looks like it might solve some of the issues with handling this kind of data...)

Reply:
Alright, sent an invitation email for our slack!